java/arcs/core/data/expression/Serializer.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+@file:Suppress("UNCHECKED_CAST")
+
+package arcs.core.data.expression
+
+import arcs.core.data.expression.Expression.BinaryExpression
+import arcs.core.data.expression.Expression.BinaryOp
+import arcs.core.data.expression.Expression.UnaryExpression
+import arcs.core.data.expression.Expression.UnaryOp
+import arcs.core.util.Json
+import arcs.core.util.JsonValue
+import arcs.core.util.JsonValue.JsonBoolean
+import arcs.core.util.JsonValue.JsonNumber
+import arcs.core.util.JsonValue.JsonObject
+import arcs.core.util.JsonValue.JsonString
+import arcs.core.util.JsonVisitor
+import arcs.core.util.ParseResult
+
+/** Traverses a tree of [Expression] objects, serializing it into a JSON format. */
+class ExpressionSerializer() :
+    Expression.Visitor<JsonValue<*>> {
+
+    override fun <E, T> visit(expr: UnaryExpression<E, T>) =
+        JsonObject(
+            mapOf(
+                "type" to JsonString(expr.op.token),
```

Comment:
`"type"` has all the type system connotations, maybe `"op"` ? I don't feel very strongly though.